framework/validators/EmailValidator.php

```diff
@@ -77,7 +77,21 @@ protected function validateValue($value)
             $valid = false;
         } else {
             if ($this->enableIDN) {
-                $matches['local'] = $this->idnToAscii($matches['local']);
+                // https://github.com/yiisoft/yii2/issues/18585
+                $localIDNTest = $this->idnToAscii($matches['local']);
```

Comment:
This method now has too many nested conditions.

Could you extract the whole branch to a separate method? Somehow like this:

```suggestion
    $value = $this->validateIdnValue($matches);
```


```php
private function validateIdnValue(&$matches)
{
    $localIDNTest = $this->idnToAscii($matches['local']);

    if ($localIDNTest === false) {
	$newPatternExploded = explode('@', $this->pattern, 2);
	$newPattern = $newPatternExploded[0] . "$/";
	$newFullPatternExploded = explode('@', $this->fullPattern);
	$newFullPattern = $newFullPatternExploded[0] . '@' . $newFullPatternExploded[1]  ."$/";
	if (!preg_match($newPattern, $matches['local']) || ($this->allowName && preg_match($newFullPattern, $matches['local']))) {
	    $matches['local'] = $localIDNTest;
	}
    } else {
	$matches['local'] = $localIDNTest;
    }

    $matches['domain'] = $this->idnToAscii($matches['domain']);
    return $matches['name'] . $matches['open'] . $matches['local'] . '@' . $matches['domain'] . $matches['close'];
}
```

Reply:
Yes, I will refactor it and resubmit it.

Reply:
Refactored to keep it more consistent with how it was originally implemented.